types/metro-resolver/errors.d.ts

```diff
@@ -0,0 +1,37 @@
+import { FileAndDirCandidates, FileCandidates } from './types';
+
+export class FailedToResolvePathError extends Error {
+    candidates: FileAndDirCandidates;
+}
+
+export class FailedToResolveNameError extends Error {
+    dirPaths: ReadonlyArray<string>;
+    extraPaths: ReadonlyArray<string>;
+}
+
+export class InvalidPackageError extends Error {
+    /**
+     * The file candidates we tried to find to resolve the `main` field of the
+     * package. Ex. `/js/foo/beep(.js|.json)?` if `main` is specifying `./beep`
+     * as the entry point.
+     */
+    fileCandidates: FileCandidates;
+
+    /**
+     * The 'index' file candidates we tried to find to resolve the `main` field of
+     * the package. Ex. `/js/foo/beep/index(.js|.json)?` if `main` is specifying
+     * `./beep` as the entry point.
+     */
+    indexCandidates: FileCandidates;
+
+    /**
+     * The module path prefix we where trying to resolve. For example './beep'.
+     */
+    mainPrefixPath: string;
+
+    /**
+     * Full path the package we were trying to resolve.
+     * Ex. `/js/foo/package.json`.
+     */
+    packageJsonPath: string;
+}
```

Comment:
```suggestion

    constructor(opts: Readonly<{
        fileCandidates: FileCandidates;
        indexCandidates: FileCandidates;
        mainPrefixPath: string;
        packageJsonPath: string;
    }>);
}
```

Reply:
@tido64 do I need to change `InvalidPackageError.fileCandidates` from `FileCandidates` to `Readonly<FileCandidates>` as well?

Reply:
I don't think so. Upstream source doesn't have the readonly specifier:
https://github.com/facebook/metro/blob/d1f3629dc75a46518a0e44d5bba05c85e887a157/packages/metro-resolver/src/InvalidPackageError.js#L24

Reply:
Oh, yeah, right